File: 0394-decode-string/0394-decode-string.cpp
class Solution {
public:
    string decodeString(string s) {
        stack<int> numStack;
        stack<string> strStack;
        int currentNum = 0;
        string currentStr;
        
        for(char ch: s){
            if(isdigit(ch)){
                currentNum = currentNum * 10 + (ch - '0');
            }
            else if(ch == '['){
                numStack.push(currentNum);
                strStack.push(currentStr);
                
                currentNum = 0;
                currentStr.clear();
            }
            else if(ch == ']'){
                int num = numStack.top();
                numStack.pop();
                string decodedStr = strStack.top();
                strStack.pop();
                
                while(num > 0){
                    decodedStr = decodedStr + currentStr;
                    num--;
                }
                currentStr = decodedStr;
            }
            else{
                currentStr = currentStr + ch;
            }
        }
        return currentStr;
    }
};

File: 0394-decode-string/0394-decode-string.java
class Solution {
    public String decodeString(String s) {
        Stack<Integer> numStack = new Stack<>();
        Stack<StringBuilder> strStack = new Stack<>();
        int currentNum = 0;
        StringBuilder currentStr = new StringBuilder();
        
        for(char ch: s.toCharArray()){
            if(Character.isDigit(ch)){
                currentNum = currentNum * 10 + (ch - '0');
            }
            else if(ch == '['){
                numStack.add(currentNum);
                strStack.add(currentStr);
                
                currentNum = 0;
                currentStr = new StringBuilder();
            }
            else if(ch == ']'){
                int num = numStack.peek();
                numStack.pop();
                StringBuilder decodedStr = strStack.peek();
                strStack.pop();
                
                while(num > 0){
                    decodedStr.append(currentStr);
                    num--;
                }
                currentStr = decodedStr;
            }
            else{
                currentStr.append(ch);
            }
        }
        return currentStr.toString();
    }
}